### ISSUE TITLE
pytest failures due to decompression of content

### ISSUE THREAD

**Issue by masselstine** [pulp/pulp_deb#742]
**Version**.
main for both pulpcore and pulp_deb, installed via pulp-installer into pulp3-source-fedora36 box.

**Describe the bug**
When running pytests

pytest -v -r sx --color=yes --pyargs pulp_deb.tests.functional -m 'parallel' -k test_download_content

you will find errors such as

Apr 04 00:55:56 pulp3-source-fedora36.example.com pulpcore-worker[37598]: Giving up download_wrapper(...) after 4 tries (pulpcore.exceptions.validation.DigestValidationError: A file located at the url http://127.0.0.1:40017/debian/dists/ragnarok/asgard/binary-ppc64/Packages.gz failed validation due to checksum. Expected 'ce2c03dc7908e4b2e71abc16dde628c5', Actual '343607ad3a5cca1d7510ab7aa31a14fa') 

These error are only observed on files which are *.gz. The discussion on aiohttp on issue #4462 appears to be related. NOTE that hacking the 'url' for a different server does not have this issue. For example a ruby oneliner: ruby3.0 -run -e httpd pulp_deb/tests/functional/data -p 8081

**To Reproduce**
Run pytest per above

**Expected behavior**
No checksum mismatch and *.gz content properly downloaded and account for in http://server.name/pulp/api/v3/content/

**Additional context**
Since we have gzip and raw content represented in the Release file in the test data we can see that this is an issue of content being compressed or decompressed as the checksums are swapped. The test itself is marked as passed since the *.gz content is not vital to the publishing of a repository. Attempting to force ClientSession to have auto_decompress=False changes the problem as the non *.gz files will start to fail due to checksum mismatch, most likely due to the content being compress in transit.

**Comment by quba42** [comment]
It is fairly critical we find a solution for this. It is a bit of a head scratcher though, since it seems to be a problem with the library (https://github.com/aio-libs/aiohttp/issues/4462). The fact that it only appears to affect the tests, makes me think it might be some configuration issue with the way we serve the local test repos?

**Comment by masselstine** [comment]
@quba42 I agree with the criticality, thus why I raised this as a separate Issue apart from the source work where I happened to first observe this. I spent quite a few hours on this yesterday. This isn't reflected by additional details captured here, as more than anything it highlighted that I have lots to learn about http. I hope that other contributors can skip the learning phase and go right to a solution.

**Comment by daviddavis** [comment]
> It is fairly critical we find a solution for this. It is a bit of a head scratcher though, since it seems to be a problem with the library (https://github.com/aio-libs/aiohttp/issues/4462).

I'm not entirely convinced this is a problem with aiohttp. Pulp [raised a similar bug that was ultimately closed](https://github.com/aio-libs/aiohttp/issues/3182#issuecomment-412634735). 

Since this seems to be happening only in the test environment, it leads me to believe that perhaps it's a problem with the test fixture server. I'm wondering if this bug should be moved to pulpcore since I am pretty certain that pulp_deb is just reusing pulpcore's downloaders without altering them, correct?

**Comment by masselstine** [comment]
> > It is fairly critical we find a solution for this. It is a bit of a head scratcher though, since it seems to be a problem with the library ([aio-libs/aiohttp#4462](https://github.com/aio-libs/aiohttp/issues/4462)).
> 
> I'm not entirely convinced this is a problem with aiohttp. Pulp [raised a similar bug that was ultimately closed](https://github.com/aio-libs/aiohttp/issues/3182#issuecomment-412634735).
> 
> Since this seems to be happening only in the test environment, it leads me to believe that perhaps it's a problem with the test fixture server. I'm wondering if this bug should be moved to pulpcore since I am pretty certain that pulp_deb is just reusing pulpcore's downloaders without altering them, correct?

This was one direction I was digging in yesterday and yes we just use the downloader as-is. The interesting thing is that Accept-Encoding set to 'identity' should be a way to resolve this, but it doesn't. I suspect the server is simply rejecting/ignoring this but I have yet to determine how to prove it. I was going to install tcpdump to be able to better monitor the interactions as part of my issue is that I just don't have the necessary data that I would need to proceed.

**Comment by masselstine** [comment]
Using tcpdump on the lo device while running the test we find that Accept-Encoding is set to 'gzip, deflate' for these *.gz files

(tcpdump -i lo -s 0 -A 'tcp[((tcp[12:1] & 0xf0) >> 2):4] = 0x47455420')
```
15:07:08.609863 IP localhost.40016 > localhost.43029: Flags [P.], seq 1126632491:1126632723, ack 362537241, win 512, options [nop,nop,TS val 2607361702 ecr 2607361700], length 232                                                                                           E...{.@.@..Z.........P..C'.+...............
.i...i..GET /debian/dists/ragnarok/jotunheimr/binary-armeb/Packages.gz HTTP/1.1
Host: 127.0.0.1:43029
User-Agent: pulpcore/3.24.0.dev0 (cpython 3.10.10-final0, Linux x86_64) (aiohttp 3.8.4)
Accept: */*
Accept-Encoding: gzip, deflate
```

I can also confirm that my attempts to force Accept-Encoding to "identity" have failed. I do believe that this is the eventual fix. As long as Accept-Encoding includes 'gzip' the server will take the *.gz files, forgo compression, deliver them to the client side where aiohttp will dutifully decompress the body and we will get the uncompressed checksum and the checksum mismatch.

**Comment by masselstine** [comment]
I lost a few days due to a later winter storm causing power outages for my equipment and the Easter long weekend but back at this today.

It appears that the server is ignoring the Accept-Encoding in the request header, though more investigation is required.
```
[vagrant@pulp3-source-fedora36 devel]$ wget -S --header="Accept-Encoding: identity"  http://127.0.0.1:56237/debian/dists/ragnarok/jotunheimr/binary-armeb/Packages.gz
--2023-04-12 19:42:46--  http://127.0.0.1:56237/debian/dists/ragnarok/jotunheimr/binary-armeb/Packages.gz
Connecting to 127.0.0.1:56237... connected.
HTTP request sent, awaiting response...
  HTTP/1.1 200 OK
  Content-Type: application/octet-stream
  Content-Encoding: gzip
  Etag: "1750f481b8f1ae00-178"
  Last-Modified: Wed, 29 Mar 2023 17:38:19 GMT
  Content-Length: 376                                                                                                                                                                                                                                                           Accept-Ranges: bytes
  Date: Wed, 12 Apr 2023 19:42:46 GMT
  Server: Python/3.10 aiohttp/3.8.4
Length: 376 [application/octet-stream]                                                                                                                                                                                                                                        Saving to: ‘Packages.gz’

Packages.gz                                                         100%[=================================================================================================================================================================>]     376  --.-KB/s    in 0s

2023-04-12 19:42:46 (29.5 MB/s) - ‘Packages.gz’ saved [376/376]
```
If the server doesn't abide by the Accept-Encoding then the only way to ensure the decompression doesn't happen is with aiohttp auto_decompress=False, but the test fixtures don't make this easy to configure.

**Comment by masselstine** [comment]
More evidence against the test fixtures server. Here are the response headers with Accept-Encoding of 'identity' and 'gzip'. NOTE the Content-Length is the same
```
[vagrant@pulp3-source-fedora36 devel]$ curl -I -H 'Accept-Encoding: identity'   http://127.0.0.1:57157/debian/dists/nosuite/asgard/binary-ppc64/Packages.gz
HTTP/1.1 200 OK
Content-Type: application/octet-stream
Content-Encoding: gzip
Etag: "1750f481b8f1ae00-363"
Last-Modified: Wed, 29 Mar 2023 17:38:19 GMT
Content-Length: 867
Accept-Ranges: bytes
Date: Thu, 13 Apr 2023 19:22:22 GMT
Server: Python/3.10 aiohttp/3.8.4

[vagrant@pulp3-source-fedora36 devel]$ curl -I -H 'Accept-Encoding: gzip'   http://127.0.0.1:57157/debian/dists/nosuite/asgard/binary-ppc64/Packages.gz
HTTP/1.1 200 OK
Content-Type: application/octet-stream
Content-Encoding: gzip
Etag: "1750f481b8f1ae00-363"
Last-Modified: Wed, 29 Mar 2023 17:38:19 GMT
Content-Length: 867
Accept-Ranges: bytes
Date: Thu, 13 Apr 2023 19:22:30 GMT
Server: Python/3.10 aiohttp/3.8.4
```
Doing a similar test using a test server I happened to find. NOTE that the size differs
```
[vagrant@pulp3-source-fedora36 devel]$ curl -I -H 'Accept-Encoding: identity' https://httpbin.org/gzip
HTTP/2 200
date: Thu, 13 Apr 2023 19:27:43 GMT
content-type: application/json
content-length: 216
server: gunicorn/19.9.0
content-encoding: gzip
access-control-allow-origin: *
access-control-allow-credentials: true

[vagrant@pulp3-source-fedora36 devel]$ curl -I -H 'Accept-Encoding: gzip' https://httpbin.org/gzip
HTTP/2 200
date: Thu, 13 Apr 2023 19:27:48 GMT
content-type: application/json
content-length: 213
server: gunicorn/19.9.0
content-encoding: gzip
access-control-allow-origin: *
access-control-allow-credentials: true
```
Again, I will preface my findings that my interpretation may be off from what the standards state, but it does appear that the fixtures server is incomplete. I plan on looking at the server code for any hints.

**Comment by daviddavis** [comment]
It seems to me that the `content-encoding: gzip` header on the response is incorrect. It's telling clients like Pulp's downloader that the original payload (Packages.gz) has been gzipped by the server and that they ought to decompress it. But this is wrong.

As a comparison, if I make a request for a Packages.gz file from fixtures.pulpproject.org, there is no content-encoding header in the response which I think is correct (and so the Downloader will not decompress it):

```
$ curl -I https://fixtures.pulpproject.org/debian/dists/ginnungagap/asgard/binary-armeb/Packages.gz
HTTP/2 200
server: nginx/1.14.1
date: Thu, 13 Apr 2023 20:10:37 GMT
content-type: application/octet-stream
content-length: 20
last-modified: Thu, 26 Jan 2023 14:18:52 GMT
etag: "63d28bcc-14"
accept-ranges: bytes
set-cookie: dd6ec0a439e94ec1afa8091bd8cb5f4f=770310614ac1e6d34d762aaa8400f5d6; path=/; HttpOnly; Secure; SameSite=None
cache-control: private                                                                                                                              
```

I think this is a bug in the test fixtures server.

**Comment by masselstine** [comment]
@daviddavis yes. Looks like aiohttp server will use mimetypes to set encoding, which then shows up as "content-encoding: gzip" in the response header, which then the aiohttp client will take as a hint to decompress the message body, which then causes the checksum mismatch.

The code of interest is here:
https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_fileresponse.py#L171

At this point I have a good feeling for the issue and can think of several fixes but need to explore some more to determine the *right* fix.

**Comment by daviddavis** [comment]
Is aiohttp what the fixture server is using? To quote [Mozilla's developer docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding), "The Content-Encoding [representation header](https://developer.mozilla.org/en-US/docs/Glossary/Representation_header) lists any encodings that have been applied to the representation (message payload), and in what order." It doesn't sound like aiohttp has applied any encodings to the content since the content is 867 bytes. So to me it sounds like a bug in aiohttp where it's incorrectly reporting the content-encoding.

**Comment by masselstine** [comment]
I have a fix. It is against pulp-smash. I am about to write up the issue and prepare and submit a pull request.

**Comment by masselstine** [comment]
@daviddavis I have created pulp/pulp-smash#1310 it would be great if you could try things out to confirm. You should be able to run test_download_content even without the source changes and observe the logs to ensure no missing checksums. The Packages.gz should also be shown when asking for the pulp_deb content.

**Comment by ggainey** [comment]
We need to move this issue to pulpcore so the fix applied to pulp-smash can be made here :+1: 
https://github.com/pulp/pulpcore/blob/0dc379742e6d51ce9b9952786af8d94e0c31001c/pulpcore/tests/functional/utils.py#L51

**Comment by dralley** [comment]
>So to me it sounds like a bug in aiohttp where it's incorrectly reporting the content-encoding.

I agree.  If the resource being downloaded is already a gzip file, content-encoding should not be set to gzip.

You can approach that in two ways:

* wget and other tools ignore content-encoding: gzip when the resource is clearly meant to be a gzip file - so that can be done client side
* or do the same hack server-side to prevent clients that know of `content-encoding: gzip` and try to decompress it from doing so

**Comment by quba42** [comment]
@masselstine Has the merge of https://github.com/pulp/pulp-smash/pull/1310 fixed this problem for your development state, or are you still experiencing issues?

If so, can we close this issue?

**Comment by masselstine** [comment]
@quba42  it should be fine to close. I need to update my vagrant box to validate things. I will try to do this over the next few days.

**Comment by masselstine** [comment]
This is confirmed fixed by pulp/pulp-smash#1310. Running
`pytest -v -r sx --color=yes --pyargs pulp_deb.tests.functional -m 'parallel' -k test_download_content`
while monitoring via **pjournal** no longer logs any checksum mismatch issues. Additionally the Packages.gz content is now found when reviewing content via the "/pulp/api/v3/content/" API.
